crates/autopilot/src/limit_order_quoter.rs

```diff
@@ -0,0 +1,94 @@
+use crate::database::Postgres;
+use anyhow::Result;
+use chrono::Duration;
+use model::{
+    order::OrderKind,
+    quote::{default_verification_gas_limit, OrderQuoteSide, QuoteSigningScheme, SellAmount},
+    signature::Signature,
+};
+use shared::order_quoting::{OrderQuoting, QuoteParameters};
+use std::sync::Arc;
+
+/// Background task which quotes all limit orders and sets the surplus_fee for each one
+/// to the fee returned by the quoting process. If quoting fails, the corresponding
+/// order is skipped.
+pub struct LimitOrderQuoter {
+    pub limit_order_age: Duration,
+    pub loop_delay: std::time::Duration,
+    pub quoter: Arc<dyn OrderQuoting>,
+    pub database: Postgres,
+}
+
+impl LimitOrderQuoter {
+    pub fn spawn(self) {
+        tokio::spawn(async move {
+            loop {
+                if let Err(err) = self.update().await {
+                    tracing::error!(?err, "failed to update limit order surplus");
+                }
+                tokio::time::sleep(self.loop_delay).await;
+            }
+        });
+    }
+
+    async fn update(&self) -> Result<()> {
+        let orders = self
+            .database
+            .limit_orders_with_outdated_fees(self.limit_order_age)
```

Comment:
That DB query should probably be limited in the number of orders it returns. In a degenerate case where we for some reason have a lot of limit orders with outdated fees it could be problematic to return all of them because of the memory use. Also would have a lot of orders here which could make the update loop take a long time. So long that some orders in the vector have already been settled or expired so getting a fee for them isn't useful.

Reply:
Maybe 100? I haven't thought about it in detail but it feels like it probably doesn't make sense to try more than that in one loop.

Reply:
With the current code, whether you limit 10 or 100 or 1000 orders is going to make little difference in terms of quotes per second because most of the time is going to be spent waiting for the quoting (I assume). You could imagine even limiting to 1 since there is no parallelism.
One problem there is that we always sleep after doing the update, which is probably not right if we have a limit. We should only sleep if there was no work to be done. If there was work, we can immediately go again because then there is likely more work.